### DISCUSSION TITLE
Development on Taichi has halted :-(    Any recommended alternatives from the community?

### DISCUSSION THREAD

**Discussion by bcolloran** [taichi-dev/taichi#8506] [General]
I notice [on the Insights > Contributors page](https://github.com/taichi-dev/taichi/graphs/contributors?from=2016-11-20&to=2024-04-02&type=c) for this repo that development on Taichi seems to have effectively stopped around mid August of last year. Oh no! :-( 

I'm a huge fan of Taichi, have been for several years now. I have been inactive recently, but was looking forward to jumping back into Taichi and seeing what is new. I'm really hoping this amazing project is still moving ahead!! <3

What does the future of Taichi hold?

**Comment by bcolloran** [comment]
I'm bummed that there has been no official response from the Taichi team here, but even more bummed to report that Taichi is indeed just in maintenance/bugfix mode and no longer under active development (according to @bobcao3 [in this discord chat](https://discord.com/channels/1072391601097289748/1072391601902592032/1221584783654518906)). I just noticed this in Discord, and thought I'd share that here for ppl like myself who visit the Taichi on Github but don't really use Discord much.

I guess it's not surprising given the the commits for the last 8 months, but it's really too bad. It's an amazing library, and I was really rooting for the Taichi team to be able to monetize their hard work while supporting this great open source tool. Best of luck to the team in their commercial product endeavors.

In my experience, when a company stops developing the project they have initiated, the project dies and the code rots... I was about to start a big project using Taichi, but I'm rethinking that now. I'm curious if folks have recommendations for other GPU compute libraries. My objectives would be:
- High level
- but still performant
- and not reliant on CUDA (to enable deploying to a range of gpu devices/architectures)

It looks like maybe WGPU/webgpu is next best bet if I don't want to hassle with super low level Vulkan code, but WGPU is still lower level than I'd prefer. Any other suggestions?

**Comment by jialing95** [comment]
Hi, thanks for raising this concern. I was also worried about Taichi's future after seeing this title, cuz I just returned to Taichi after long absence of use.

I read the chat history between you and bobcao3 on Discord. He said they are using Taichi in their internal commercial projects. It doesn't mean they will abandon Taichi gradually, in contrast, they are building products based on it, as the team need the profit to support I think. The new release v.1.7.1 was published on 4/17/2024. So we can keep checking the active development of Taichi for the following months. 

Cheers!

**Reply by bcolloran** [reply]
interesting @jialing95, I read that exchange differently (it was not me chatting with @bobcao3, btw) -- to me it, it sounded like they would only be adding fixes as needed when aligned with their commercial projects. And looking at the commit history for the past 8 months, it seems pretty clear that the commercial interests have diverged with those of the Taichi Project:

![image](https://github.com/taichi-dev/taichi/assets/2380975/3c72de14-96f2-4de7-b4e4-93977466447e)
https://github.com/taichi-dev/taichi/graphs/contributors

I also take it as a signal that no one actually from the Taichi organization has made any comment on this thread. The organization used to be very actively engaged with the community around Taichi, and lack of response to a thread with a concerning title like this is a strong indicator that they are no longer have time to spend with the open source Taichi project, as their strategy has changed to focus their time on other commercial offerings.

Again, I have no hard feelings towards the Taichi company. I am amazed by this project and greatly respect the Taichi team for what they have accomplished, and having run open source projects myself, I understand the challenges around commercialization needing to pivot if the project isn't generating the returns you hoped for. So I wish the Taichi team the best of luck in their future commercial endeavors.

But I think it is very clear that active development on Taichi wound down last summer, and that the project is in maintenance mode _at best_, and so I am thinking carefully about whether to use Taichi going forward, and I'm very interested in whether anyone has knowledge of alternative option. I think it will be hard to find since Taichi is so very advanced, but I'm wary of launching my own new projects based on something which no longer has energy behind it, and which the sponsoring company was not able to build a commercial strategy around and therefore pivoted away from. In my experience these are both early signs that a project may be abandoned, which makes me nervous.

**Comment by jialing95** [comment]
I sincerely wish @yuanming-hu can give us an explaination what's going on there. I started to use Taichi to do the homework recently, and was hoping to use it to build SPH models for my research..

**Comment by yuanming-hu** [comment]
Hi folks, thanks for your ongoing discussions and support. It’s heartening to see such a passionate community around Taichi Lang.

While it’s true that our active development pace has moderated, Taichi is still **very much alive and maintained**. Our shift in focus, towards building a commercial GenAI product using Taichi, is aimed at ensuring our team's sustainability and allowing us to continue supporting and developing Taichi.

Taichi was designed for high-performance physical simulation and differentiable graphics, and we are proud to have achieved a level of feature completeness and stability in these domains. Our current challenge (and opportunity) is in determining the next significant milestones for Taichi 2.0. While the future path is not fully defined, we are excited about the possibilities and welcome your thoughts on potential directions. (One possibility is more support for differentiable graphics, given GenAI, leveraging differentiable graphics, nowadays have increasing impact.)

In the near term, you can expect updates such as support for CUDA 12.4, Python 3.12, and Ubuntu 24.04, along with ongoing bug fixes. We are committed to ensuring Taichi remains reliable and up-to-date.

Taichi is still actively used within our company for various commercial projects, so it's certainly not going to be abandoned. Last week I personally wrote over 500 lines of Taichi code! We are fully committed to its future and are contemplating what new features - like enhanced automatic differentiation, additional backends, or better modularity - would best serve our community.

**Reply by jialing95** [reply]
Thanks for updating us Taichi's new features and visions. I recently noticed Taichi has been incorporated in MPM and SPH frameworks among some Chinese researchers. It's a good trend. We must keep Taichi alive and regularly maintain it. 

Hope Taichi team can achieve continuous progress as before!

**Reply by yuanming-hu** [reply]
Yeah simulaition is our base, and we are actively expanding Taichi to differentiable simulation and rendering :-)

**Reply by Paulo-21** [reply]
Hello, Thank you for your work on taichi, I would like to know if there is any news about a taichi version build with cuda 12.X ?
I tried to compiled it with cuda 12.9 but it failed.
I think supporting a recent version of cuda could help with performance with my recent nvidia card.
I will realy appreciate an update about it :)
Best Regards

**Comment by parduino** [comment]
Hi,

We have been using taichi in several MPM and fluids projects.
We use it in win, OSX, and ubuntu distributions and installed it in TACC (Texas Austing Comuputing Center) HPC systems. 
It works really well.

We were recently trying to port it to the new TACC system called Vista that is based on ARM infrastructure.
We have been trying to build it from source but we are getting the error:

runtime.cpp:1160:18: error: invalid input constraint 'l' in asm
 1160 |                : "l"(value), "r"(mask));
 
Searching online we get:

This is a deep source code error that you won't be able to fix with build configuration changes. It indicates that a piece of the Taichi source code is incompatible with the ARM64 architecture of the Grace-Hopper CPU.

The Core Problem: Invalid Assembly Code
The error invalid input constraint 'l' in asm is coming from a block of inline assembly code within Taichi's source file runtime.cpp.

In C++, inline assembly allows developers to write low-level code that speaks directly to the CPU. To do this, they use "constraints" to tell the compiler how to handle variables. The error message is telling you that the constraint 'l' is not valid on your machine.

That's because the 'l' constraint is specific to the older 32-bit ARM architecture (AArch32). It is not supported on the modern 64-bit ARM architecture (AArch64) used by the NVIDIA Grace CPU.

This is a bug in the Taichi source code where a piece of 32-bit-specific code is being incorrectly compiled on a 64-bit system.

The Only Solution: Report the Bug
You have successfully navigated a very complex build process, but this is a roadblock that requires a code change from the Taichi developers.

The best course of action is to open an issue on the Taichi GitHub page.

Any plans to make taichi working in ARM systems??

Thanks

**Reply by houkensjtu** [reply]
Hi @parduino , there have been some efforts in the community: https://github.com/taichi-dev/taichi/issues/7916 . Have you tried the workaround [here](https://github.com/taichi-dev/taichi/issues/7916#issuecomment-1535641223)?
I used this workaround back in 2023 and it did work out on a Jetson Nano board.
